### PR TITLE
Full switch to glvnd

### PR DESCRIPTION
**Pull request by xtraeme**
This branch fully switches your system to use glvnd, I've been running this for a while already on my x86_64-musl system with no issues. Packages do not even need to be recompiled because libGL/EGL/GLES dependencies are not changed.

The libglvnd package now provides libGL/EGL/GLES replacing old mesa/nvidia-libs packages.

Based on work by Stefano Ragni via #16497 

Closes #12688 
Closes #16927 
Closes #16497

### CODE REVIEW

**Comment by xtraeme** [comment]
Current status

[x] - x86_64-musl mesa
[ ] - x86_64 with mesa
[ ] - x86_64 with nvidia
[ ] - x86_64 with nvidia390
[ ] - x86_64 with nvidia340
[ ] - i686 with mesa

**Comment by xtraeme** [comment]
@st3r4g @abenson please test!

**Comment by st3r4g** [comment]
After a quick look, I like the changes introduced here. I'll test `x86_64-musl mesa` in the following days and `x86_64 with mesa/nvidia390` (Optimus laptop) next weekend. One thing I'm not sure is if the `libglx-nvidia.so` symlink in `nvidia390` is enough to make X11 pick up the right GLX, but I'll test and we'll see.

**Comment by xtraeme** [comment]
@st3r4g the libglx-nvidia.so symlink I think it's even unnecessary. the xorg-server code should now check for possible candidates at /usr/lib/xorg/modules/extensions automagically (guess this even before looking at the code).

**Comment by st3r4g** [comment]
Some comments (the list could grow):
- `nvidia340` does not support `glvnd`, so it must be changed to conflict/replace `libglvnd` (same mechanism `libGL` and `nvidia-libs` are using now). Use case is people with just one (old) nvidia gpu. Same thing for `catalyst` and `sunxi-mali`.
- now it would make sense to rename `MesaLib-devel` into `mesa-devel`

Just installed ` x86_64-musl mesa` (everything but nvidia), which is running fine on my Intel Graphics laptop, tested both X (GLX) and Wayland (EGL).

File: srcpkgs/libglvnd/files/LICENSE [hunk 1]
```diff
@@ -0,0 +1,26 @@
+Copyright (c) 2013, NVIDIA CORPORATION.
```
Comment (Johnnynator) [inline srcpkgs/libglvnd/files/LICENSE]: Pls don't vendor licenses in `void-packages`

Comment (st3r4g) [reply srcpkgs/libglvnd/files/LICENSE]: Isn't that basically a MIT license? [here](https://gitlab.freedesktop.org/glvnd/libglvnd/merge_requests/23) sounds like it should be included. Do we need to include also the whole 'Acknowledgements' section of the `README.md`?

File: srcpkgs/libglvnd/template [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+pkgname=libglvnd
+version=1.2.0
+revision=1
+wrksrc="libglvnd-v${version}"
+build_style=gnu-configure
+hostmakedepends="autoconf libtool automake pkg-config python"
+makedepends="libXext-devel libX11-devel xorgproto"
+short_desc="GL Vendor-Neutral Dispatch library"
+maintainer="Stefano Ragni <st3r4g@protonmail.com>"
+license="custom:NVIDIA"
```
Comment (st3r4g) [inline srcpkgs/libglvnd/template]: Isn't the license MIT?

Comment (aeadio) [reply srcpkgs/libglvnd/template]: This is a modified MIT with additional restrictions. Namely,

> Any additions, deletions, or changes to the original source files
> must be clearly indicated in accompanying documentation.
> 
> If only executable code is distributed, then the accompanying
> documentation must state that "this software is based in part on the
> work of the Khronos Group."

I don't think it's OK to just mark as MIT. I'm not sure what verbiage Void typically uses here -- maybe "Modified MIT" would be applicable. Custom by Nvidia is not inaccurate -- nobody stated it's proprietary.

**Comment by xtraeme** [comment]
I'm interested in testers with nvidia hw. I'd like to merge this before 2020 :-)

**Comment by grand0rbiter** [comment]
I have a modern laptop with nvidia and intel gpu, but i'm a total newbie about doing this.

What do i need to do?

**Comment by aeadio** [comment]
> I'm interested in testers with nvidia hw. I'd like to merge this before 2020 :-)

I've got an Optimus laptop (XPS 15) and a desktop with Nvidia dGPU + Intel iGPU. Won't have time to test for a week or two, but if nobody's stepped forward by then, happy to. 

Are there any specific test cases you're looking for? Any expected breakage or edge cases expected?

**Comment by st3r4g** [comment]
> the xorg-server code should now check for possible candidates at /usr/lib/xorg/modules/extensions automagically (guess this even before looking at the code).

@xtraeme It's not that easy, at least for the `nvidia390` case. Today I could test again on my Optimus laptop, and I could never get X to load the nvidia glx without symlinking `libglx.so.390.132 -> libglx.so`. So removing that will probably break GL for people running on a single nvidia card with `nvidia390`. `nvidia390` will probably be the hardest to adapt correctly for both the single gpu and Optimus cases, also because its Optimus support is half-assed.

@aeadio that would be amazing, I think we shouldn't rush and take all the necessary time to test possible configurations, as the changes introduced here are not trivial.
>Are there any specific test cases you're looking for?

Yes, the main purpose of this PR is to allow this:
http://us.download.nvidia.com/XFree86/Linux-x86_64/440.36/README/primerenderoffload.html (see also my previous PR for general info about Optimus).
If would be great if you or some other tester could achieve GPU offloading by the mechanism described there.

**Comment by xtraeme** [comment]
Ok, I see. We are missing this file from Arch: https://git.archlinux.org/svntogit/packages.git/tree/trunk/nvidia-drm-outputclass.conf?h=packages/nvidia-390xx-utils

@st3r4g please test ^

**Comment by st3r4g** [comment]
> We are missing this file from Arch: https://git.archlinux.org/svntogit/packages.git/tree/trunk/nvidia-drm-outputclass.conf?h=packages/nvidia-390xx-utils

Yes, as I was saying in previous PR, we need to:
1. Install `libglx.so.${version}` in a different directory than xorg's glx. Arch is using `/usr/lib/nvidia/xorg/` but that is arbitrary, I proposed the path `/usr/lib/xorg/modules-nvidia`
2. Append this path to `nvidia-drm-outputclass.conf` in the following way:
```
...
MatchDriver "nvidia-drm"
Driver "nvidia"
ModulePath "/usr/lib/xorg/modules-nvidia" # <-- Add this here
EndSection
```

That will load nvidia's glx by default. Notice that `nvidia-drm-outputclass.conf` is already provided by nvidia, and Arch is modifying it to fit their needs. I'd rather not copy their whole custom configuration but only implement the line which loads the correct glx.

**Comment by xtraeme** [comment]
ok, finally got this all working ootb. Tested with integrated intel gpu and nvidia gtx 960. Both setups worked with no additional changes.

See attached screenshots.
IMG_20191207_133749
IMG_20191207_124747
IMG_20191207_123910

**Comment by xtraeme** [comment]
@st3r4g let me update nvidia390 now for you to test. You can build quickly an iso image for testing via void-mklive... this is how I tested it with all possible setups.

**Comment by xtraeme** [comment]
@st3r4g it's ready now. Easy way for testing:

```
$ ./xbps-src -t pkg nvidia390 xorg-server mesa
$ cd ~/void-mklive
$ sed -e 's,xorg-video-drivers,,g' build-x86-images.sh.in
$ make
$ sudo ./build-x86-images.sh -b mate -r /path/to/hostdir/binpkgs/glvnd -r /path/to/hostdir/binpkgs/glvnd/nonfree -r https://alpha.de.repo.voidlinux.org/current/nonfree -p "mesa-intel-dri nvidia"
```
and then write the resulting iso image in a pendrive and enjoy.

It's very important that **xf86-video-intel** is not installed, that's why `xorg-video-drivers` is removed.

Enjoy

**Comment by st3r4g** [comment]
ok, but I won't get the same results as yours because the legacy `nvidia390` doesn't support the new mechanism. Best I can do with this driver is https://download.nvidia.com/XFree86/Linux-x86_64/390.132/README/randr14.html, which is choosing the GPU at the start of X session, and it's impossible to achieve purely ootb.

**Comment by xtraeme** [comment]
Even if you can't use PRIME render offloading, it should allow you to use nvidia390 and mesa without any file conflict. This is not a showstopper issue.

**Comment by st3r4g** [comment]
Sure, it's still an improvement. I tested and now when `nvidia390` is installed the glx defaults to the nvidia one, which is good for nvidia gpu only users. To run X on intel instead, one has to override the ModulePath in some way, maybe we can put this in the manual in a section specifically for Optimus users with 390.

**Comment by xtraeme** [comment]
@st3r4g please paste Xorg.0.log

**Comment by st3r4g** [comment]
In which configuration? ootb Optimus with `nvidia390` doesn't load GL correctly, one has to use the `/etc/X11/xorg.conf` described in link above and also modify `.xinitrc`

[ootb.txt](https://github.com/void-linux/void-packages/files/3935179/ootb.txt)
[nvidia.txt](https://github.com/void-linux/void-packages/files/3935180/nvidia.txt) (after following nvidia instructions)
[intel.txt](https://github.com/void-linux/void-packages/files/3935181/intel.txt) (remove the `/etc/X11/xorg.conf` from nvidia instructions and comment the ModulePath in `nvidia-drm-outputclass.conf` [I know commenting stuff is not viable, I have still to find a better way to override that]).

**Comment by xtraeme** [comment]
ok let me know what we are missing for nvidia390. nvidia and mesa are ready and tested.

**Comment by st3r4g** [comment]
I do not think we can make `nvidia390` work ootb for both the single nvidia gpu case and optimus case

**Comment by xtraeme** [comment]
@st3r4g this patch (which we don't ship) should do the trick for you: https://git.archlinux.org/svntogit/packages.git/tree/trunk/xserver-autobind-hotplug.patch?h=packages/xorg-server

**Comment by xtraeme** [comment]
I also think we should apply this patch, seems to be useful.

**Comment by st3r4g** [comment]
I'm against introducing more features in this PR, which is already quite complex. `nvidia390` is already in a better shape and will save users to do all [this](https://www.ifnull.org/articles/void_optimus_nvidia_intel/) because mesa and nvidia do not conflict anymore as packages. Maybe we can improve nvidia390's Optimus later in another PR.

**Comment by xtraeme** [comment]
So with this PR in its current state, nvidia390 as primary GPU works? intel as primary GPU works?

**Comment by xtraeme** [comment]
I'm going to clean this up and merge.

File: srcpkgs/libglvnd/files/LICENSE (hunk 1 above)
Comment (xtraeme) [reply srcpkgs/libglvnd/files/LICENSE]: so what's the deal with this?

File: srcpkgs/libglvnd/template (hunk 2 above)
Comment (xtraeme) [reply srcpkgs/libglvnd/template]: so what's the deal?

Comment (xtraeme) [reply srcpkgs/libglvnd/template]: I set this as `custom:MIT-alike` for now

File: srcpkgs/libglvnd/files/LICENSE (hunk 1 above)
Comment (xtraeme) [reply srcpkgs/libglvnd/files/LICENSE]: We don't vendor the license anymore

**Comment by st3r4g** [comment]
> So with this PR in its current state, nvidia390 as primary GPU works? intel as primary GPU works?

On my laptop, nvidia390 as primary didn't work ootb with current packages (and won't work ootb after this). It's not wired to the screen, so it doesn't work without `xrandr --setprovideroutputsource modesetting NVIDIA-0` somewhere.

**Comment by xtraeme** [comment]
@st3r4g that's why I told you about the xorg-server patch, because this patch avoids the need to exec the xrandr command.

**Comment by st3r4g** [comment]
I'm trying that patch, but I still need a custom `xorg.conf` to make it work, and I don't know if we can ship it by default, it could break the single gpu case. Now I don't need anymore `xrandr --setprovideroutputsource modesetting NVIDIA-0` but still I need to start with Intel screen disabled and enable with `xrandr --auto`...

**Comment by xtraeme** [comment]
what's in your custom `xorg.conf` ?

**Comment by st3r4g** [comment]
It's the one in https://download.nvidia.com/XFree86/Linux-x86_64/390.132/README/randr14.html

**Comment by xtraeme** [comment]
@st3r4g for nvidia390 we should use a different nvidia-drm-outputclass.conf probably. You are the one that can test this hw. Please be my guest :-)

**Comment by xtraeme** [comment]
If we can't make this work ootb, we should make it use at least work for the single GPU case.

**Comment by st3r4g** [comment]
Don't have a single gpu setup, but I think it works there as it is, because there will be modesetting by nvidia and glx also defaults to nvidia. The problem with Optimus is that, when `nvidia390` is installed, X does by default the modesetting on Intel while loads the glx from nvidia, and this results in broken GL ootb.

**Comment by xtraeme** [comment]
but I'm interested when the NVIDIA GPU is the primary (enabled at BIOS/firmware init). Does your laptop allow this?

**Comment by st3r4g** [comment]
nope

**Comment by aeadio** [comment]
Unless I'm misunderstanding, we don't _want_ the Nvidia to be primary. That would be an exceptional case that nobody should expect ootb. You don't want the Nvidia powered on for desktop, web browser, and other simple cases. 

I'm not sure how other distros do it (if at all), but users should expect: 

- For single-GPU Intel systems, everything just works on Intel
- For single-GPU Nvidia systems, everything just works on Nvidia
- For Optimus setups, everything works on Intel by default, Nvidia is powered off by default, and there's is some mechanism (ideally a helper program) to launch applications on the Nvidia GPU. It powers up for these, then back down afterwards. 

I'm not sure to what extent the new PRIME improvements in the latest drivers automate the power up/down (if at all). We may still need something like bbswitch for this. At least the rendering part is handled by just exporting a few env variables, so it's simple enough to ship a wrapper in /bin that execs after exporting those.

**Comment by st3r4g** [comment]
@aeadio I was speaking about the nvidia390 (legacy driver) which doesn't allow true Optimus like you see in the screenshots, but only to choose the gpu for the whole X session at Xorg startup. nvidia390 doesn't actually benefit from libglvnd at all, if not for the fact that the various GL libs can be better installed without name conflicts.
As for true Optimus with the latest driver, I think we must rely on the power management of the driver. Using bbswitch manually will break things imho because X is nvidia-aware in that configuration.

**Comment by xtraeme** [comment]
I've been playing to a few games on steam with this setup. Running nicely.
Played to `The Witcher 2`, `Left 4 Dead 2` and `Life is strange`.

This means that 32bit stuff also works nicely.  This PR is now ready to be merged.

@void-linux/pkg-committers I'd like to push this tomorrow.

**Comment by xtraeme** [comment]
@q66 could please test on ppc?

**Comment by xtraeme** [comment]
Just FYI I made all tests by using a custom iso image via `void-mklive` and el-cheapo usb storage device :-)

**Comment by q66** [comment]
@xtraeme my setup is currently a bit "special" as i'm using mesa-git with llvm-git in order to get my RX 5700 XT GPU to work reliably, but I guess I can test it on my other machines. That said, I expect no issues whatsoever

**Comment by xtraeme** [comment]
I'm merging this. This worked in all possible combos I tried: glibc and musl, mesa+nvidia, only nvidia.

Let the party begin
